### PR TITLE
Feature/resend confirmation

### PR DESCRIPTION
**Pull request by chennan47**
<b>Purpose</b>
Allow user to resend email confirmation to added unconfirmed email. Solve https://github.com/CenterForOpenScience/osf.io/issues/2487.

<b>Changes</b>
screen shot 2015-04-18 at 2 52 38 pm
Now there is a resend confirmation button which will resend confirmation token to your email address.

### CODE REVIEW

**Comment by chennan47** [comment]
@sloria rdy for review.

File: website/static/js/accountSettings.js [hunk 1]
```diff
@@ -193,6 +210,21 @@ var UserProfileViewModel = oop.extend(ChangeMessageMixin, {
...
+    reSendConfirm: function(email){
+        var self = this;
+        self.changeMessage('', 'text-info');
+        bootbox.confirm({
+            title: 'Resend Email Confirmation?',
+            message: 'Are you sure that you want to resend email confirmation at ' + '<em><b>' + email.address() + '</b></em>',
+            callback: function (confirmed) {
+                if (confirmed) {
+                    self.client.update(self.profile(), email).done(function () {
+                        $osf.growl('Email confirmation resends to ', '<em>' + email.address() + '<em>', 'success');
```
Comment (samanehsan) [inline website/static/js/accountSettings.js]: Please change "resends" to "resent" here.

File: website/profile/views.py [hunk 2]
```diff
@@ -83,22 +83,51 @@ def date_or_none(date):
...
+        address = data['email']['address'].strip().lower()
+    except KeyError:
+        raise HTTPError(httplib.BAD_REQUEST)
+
+    if primary or confirmed:
+        raise HTTPError(httplib.BAD_REQUEST, data={'message_long': 'Cannnot resend confirmation for confirmed emails'})
+
+    user.add_unconfirmed_email(address)
+
+    # TODO: This setting is now named incorrectly.
```
Comment (samanehsan) [inline website/profile/views.py]: Could you clarify what you mean by this TODO? Why is the setting name incorrect?

**Comment by samanehsan** [comment]
Done with first pass :octocat:

File: website/profile/views.py (hunk 2 above)
Comment (chennan47) [inline website/profile/views.py]: This is from https://github.com/CenterForOpenScience/osf.io/blob/develop/website/profile/views.py#L151

File: website/static/js/accountSettings.js (hunk 1 above)
Comment (chennan47) [inline website/static/js/accountSettings.js]: addressed in  b4b63e1

File: website/profile/views.py (hunk 2 above)
Comment (samanehsan) [inline website/profile/views.py]: Oh, okay. I thought it was something you added in. @lyndsysimon, can you clarify what you mean in this TODO?

Comment (lyndsysimon) [inline website/profile/views.py]: It should be `CONFIRM_EMAILS` or similar, as there are now cases other than registrations where emails are confirmed (adding and changing).

This need not be done in the PR.

File: framework/auth/core.py [hunk 3]
```diff
@@ -666,6 +665,11 @@ def add_unconfirmed_email(self, email, expiration=None):
 
         validate_email(email)
 
+        # If the unconfirmed email is already present, refresh the token
+        if email in self.unconfirmed_emails:
+            self.remove_unconfirmed_email(email)
+            self.save()
```
Comment (lyndsysimon) [inline framework/auth/core.py]: Since the user is saved later in this method, can we avoid saving it here?

File: website/profile/views.py (hunk 2 above)
Comment (samanehsan) [inline website/profile/views.py]: Ok, so you mean consolidating this under a more general flag?

File: website/static/js/accountSettings.js [hunk 4]
```diff
@@ -89,13 +90,18 @@ var UserProfileClient = oop.defclass({
 
         return ret.promise();
     },
-    update: function (profile) {
+    update: function (profile, email) {
+        var self = this;
```
Comment (lyndsysimon) [inline website/static/js/accountSettings.js]: There's no need to define `self` here, as `this` is bound in context at the end of the function definition.

If you keep `self`, let's from `.bind(this)` below.

File: website/static/js/accountSettings.js [hunk 5]
```diff
@@ -193,6 +210,21 @@ var UserProfileViewModel = oop.extend(ChangeMessageMixin, {
             this.changeMessage('Email cannot be empty.', 'text-danger');
         }
     },
+    reSendConfirm: function(email){
```
Comment (lyndsysimon) [inline website/static/js/accountSettings.js]: This should be renamed `resendConfim`, or even `resendConfirmation`.

**Comment by lyndsysimon** [comment]
Second pass complete :tram:

File: framework/auth/core.py (hunk 3 above)
Comment (chennan47) [inline framework/auth/core.py]: addressed in da5feb2

File: website/profile/views.py (hunk 2 above)
Comment (lyndsysimon) [inline website/profile/views.py]: I think just changing the name would be fine. The purpose of the flag jsut doesn't match its name anymore.

File: website/static/js/accountSettings.js (hunk 5 above)
Comment (chennan47) [inline website/static/js/accountSettings.js]: address in 777d47b

File: website/static/js/accountSettings.js (hunk 4 above)
Comment (chennan47) [inline website/static/js/accountSettings.js]: address in 777d47b

**Comment by chennan47** [comment]
rdy for another pass

**Comment by samanehsan** [comment]
Ping @lyndsysimon

**Comment by sloria** [comment]
I can do next review of this.

File: framework/auth/core.py [hunk 6]
```diff
@@ -709,8 +712,6 @@ def add_unconfirmed_email(self, email, expiration=None):
 
     def remove_unconfirmed_email(self, email):
         """Remove an unconfirmed email addresses and their tokens."""
-        if email == self.username:
```
Comment (sloria) [inline framework/auth/core.py]: @chennan47 Why was this removed?

File: website/static/js/accountSettings.js [hunk 7]
```diff
@@ -193,6 +209,21 @@ var UserProfileViewModel = oop.extend(ChangeMessageMixin, {
             this.changeMessage('Email cannot be empty.', 'text-danger');
         }
     },
+    reSendConfirmation: function(email){
```
Comment (sloria) [inline website/static/js/accountSettings.js]: Minor typo: `resendConfirmation`

File: website/static/js/accountSettings.js [hunk 8]
```diff
@@ -193,6 +209,21 @@ var UserProfileViewModel = oop.extend(ChangeMessageMixin, {
...
+        var self = this;
+        self.changeMessage('', 'text-info');
+        bootbox.confirm({
+            title: 'Resend Email Confirmation?',
+            message: 'Are you sure that you want to resend email confirmation at ' + '<em><b>' + email.address() + '</b></em>',
+            callback: function (confirmed) {
+                if (confirmed) {
+                    self.client.update(self.profile(), email).done(function () {
+                        $osf.growl('Email confirmation resent to ', '<em>' + email.address() + '<em>', 'success');
```
Comment (sloria) [inline website/static/js/accountSettings.js]: The growlbox looks awkward:

osf___account_settings

It should be made consistent with the other growlbox for sending a confirmation link:

osf___account_settings

**Comment by sloria** [comment]
Looks good. Just minor changes and this is good to merge. :+1:

File: framework/auth/core.py (hunk 6 above)
Comment (chennan47) [inline framework/auth/core.py]: When this function is called, your primary email should always be verified which will never be in the unconfirmed email list. If it is, then it is wrong with other code. But this check is not necessary.

File: website/static/js/accountSettings.js (hunk 7 above)
Comment (chennan47) [inline website/static/js/accountSettings.js]: fixed in 6e6e6fe

File: website/static/js/accountSettings.js (hunk 8 above)
Comment (chennan47) [inline website/static/js/accountSettings.js]: fixed in 6e6e6fe.
Now:
screen shot 2015-06-11 at 11 01 16 am

File: website/routes.py [hunk 9]
```diff
@@ -427,13 +427,6 @@ def make_url_map(app):
         ),
 
         Rule(
-            '/resend/',
```
Comment (sloria) [inline website/routes.py]: @chennan47 Just a heads up--turns out we actually do need this route. This is used for resending confirmation emails after signing up. The "Resend" link on the CAS login page links to it.

I'll work on readding this.

Comment (lyndsysimon) [inline website/routes.py]: I just checked, this is in-place on `develop`.

Comment (sloria) [inline website/routes.py]: Yeah, I added it in a hotfix on Friday.
